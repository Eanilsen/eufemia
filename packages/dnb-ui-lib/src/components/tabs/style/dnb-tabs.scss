/*
* DNB Tabs
*
* ======================================================================== */

@import '../../../style/component/core-imports.scss';
// @import '../../../style/component/font-import.scss';
@import '../../icon/style/dnb-icon.scss';
@import '../../button/style/dnb-button.scss';
@import './_tabs.scss';
