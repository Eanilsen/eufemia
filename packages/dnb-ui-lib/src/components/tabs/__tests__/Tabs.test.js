/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  fakeAllProps,
  axeComponent,
  toJson,
  loadScss
} from '../../../core/jest/jestSetup'
import Component from '../Tabs'
// just to make sure we re-run the test in watch mode due to changes in this file
import '../style/dnb-tabs.scss'

const props = fakeAllProps(require.resolve('../Tabs'), {
  optional: true
})
delete props.render

const startup_selected_key = 'second'
const tablistData = [
  { title: 'First', key: 'first' },
  { title: 'Second', key: 'second' },
  { title: 'Third', key: 'third' }
]
const tabContentData = {
  first: <h2>First</h2>,
  second: <h2>Second</h2>,
  third: <h2>Third</h2>
}

describe('Tabs component', () => {
  const Comp = mount(
    <Component
      {...props}
      data={tablistData}
      selected_key={startup_selected_key}
    >
      {tabContentData}
    </Component>
  )

  it('have to match snapshot', () => {
    expect(toJson(Comp)).toMatchSnapshot()
  })

  it('have a "selected_key" state have to be same as prop from startup', () => {
    expect(Comp.state().selected_key).toBe(startup_selected_key)
  })

  it('should validate with ARIA rules as a tabs', async () => {
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})

describe('TabList component', () => {
  const Comp = mount(
    <Component
      {...props}
      data={tablistData}
      selected_key={startup_selected_key}
    >
      {tabContentData}
    </Component>
  )

  it('has to have the right amount of renderet components', () => {
    expect(Comp.find('button').length).toBe(tablistData.length)
    expect(Comp.find('div[role="tabpanel"]').length).toBe(1)
  })

  it('has to have the right content on a "click event"', () => {
    Comp.find('button.tab--third').simulate('click')
    expect(Comp.state().selected_key).toBe(tablistData[2].key) // get the third key
    expect(Comp.find('div[role="tabpanel"]').text()).toBe(
      mount(tabContentData.third).text()
    )
  })
})

describe('A single Tab component', () => {
  const Comp = mount(
    <Component
      {...props}
      data={tablistData}
      selected_key={startup_selected_key}
    >
      {tabContentData}
    </Component>
  )

  it('has to have the right content on a keydown "ArrowRight"', () => {
    expect(
      Comp.find('button.tab--second')
        .instance()
        .getAttribute('aria-controls')
    ).toBe('id-content-second')
    expect(Comp.find('button.tab--second').is('.selected')).toBe(true)
  })

  it('has to have the right content on a keydown "ArrowRight"', () => {
    // reset the state
    Comp.find('button.tab--second').simulate('click')
    Comp.find('div[role="tablist"]').simulate('keyDown', {
      key: 'ArrowRight',
      keyCode: 39
    })
    expect(Comp.find('div[role="tabpanel"]').text()).toBe(
      mount(tabContentData.third).text()
    )
  })
})

describe('Tabs scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(require.resolve('../style/dnb-tabs.scss'))
    expect(scss).toMatchSnapshot()
  })
})
