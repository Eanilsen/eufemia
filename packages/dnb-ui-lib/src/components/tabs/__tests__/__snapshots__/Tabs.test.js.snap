// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component have to match snapshot 1`] = `
<Tabs
  class="class"
  className="className"
  data={
    Array [
      Object {
        "key": "first",
        "title": "First",
      },
      Object {
        "key": "second",
        "title": "Second",
      },
      Object {
        "key": "third",
        "title": "Third",
      },
    ]
  }
  direction="left"
  do_set_hash="do_set_hash"
  id="id"
  label="label"
  render={null}
  selected_key="second"
  selection_key="selection_key"
>
  <Wrapper>
    <div
      className="dnb-tabs className class"
    >
      <TabsList>
        <div
          className="dnb-tabs__tabs dnb-tabs__tabs--left"
        >
          <Tabs>
            <div
              aria-label="label"
              className="dnb-tabs__tabs__tablist dnb-tab-focus"
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              role="tablist"
              tabIndex="0"
            >
              <button
                aria-selected={false}
                className="dnb-tablink dnb-no-mouse-focus tab--first "
                disabled={false}
                id="id-tab-first"
                onClick={[Function]}
                role="tab"
                tabIndex="-1"
              >
                <span
                  className="dnb-tablink-title"
                >
                  First
                </span>
                <span
                  aria-hidden={true}
                  hidden={true}
                >
                  First
                </span>
              </button>
              <button
                aria-controls="id-content-second"
                aria-selected={true}
                className="dnb-tablink dnb-no-mouse-focus tab--second selected"
                disabled={false}
                id="id-tab-second"
                onClick={[Function]}
                role="tab"
                tabIndex="-1"
              >
                <span
                  className="dnb-tablink-title"
                >
                  Second
                </span>
                <span
                  aria-hidden={true}
                  hidden={true}
                >
                  Second
                </span>
              </button>
              <button
                aria-selected={false}
                className="dnb-tablink dnb-no-mouse-focus tab--third "
                disabled={false}
                id="id-tab-third"
                onClick={[Function]}
                role="tab"
                tabIndex="-1"
              >
                <span
                  className="dnb-tablink-title"
                >
                  Third
                </span>
                <span
                  aria-hidden={true}
                  hidden={true}
                >
                  Third
                </span>
              </button>
            </div>
          </Tabs>
        </div>
      </TabsList>
      <TabContent
        id="id"
        selection_key="second"
      >
        <div
          aria-labelledby="id-tab-second"
          className="dnb-tabs__content"
          id="id-content-second"
          role="tabpanel"
        >
          <h2>
            Second
          </h2>
        </div>
      </TabContent>
    </div>
  </Wrapper>
</Tabs>
`;

exports[`Tabs scss have to match snapshot 1`] = `
"/*
* DNB Tabs
*
* ======================================================================== */
/*
 * Utilities
 * ========================================================================== */
/*
* Scopes
* ======================================================================== */
@font-face {
  font-family: 'Fedra Sans Std Book';
  src: url(\\"../../../../assets/fonts/FedraSansStd-Book.eot\\");
  src: url(\\"../../../../assets/fonts/FedraSansStd-Book.eot?#iefix\\") format(\\"embedded-opentype\\"), url(\\"../../../../assets/fonts/FedraSansStd-Book.woff\\") format(\\"woff\\"), url(\\"../../../../assets/fonts/FedraSansStd-Book.ttf\\") format(\\"truetype\\");
  font-weight: normal;
  font-style: normal; }

@font-face {
  font-family: 'Fedra Sans Std Demi';
  src: url(\\"../../../../assets/fonts/FedraSansStd-Demi.otf\\") format(\\"opentype\\"), url(\\"../../../../assets/fonts/FedraSansStd-Demi.woff\\") format(\\"woff\\"), url(\\"../../../../assets/fonts/FedraSansStd-Demi.ttf\\") format(\\"truetype\\");
  font-weight: normal;
  font-style: normal; }

@font-face {
  font-family: 'Fedra Sans Std Medium';
  src: url(\\"../../../../assets/fonts/FedraSansStd-Medium.eot\\");
  src: url(\\"../../../../assets/fonts/FedraSansStd-Medium.eot?#iefix\\") format(\\"embedded-opentype\\"), url(\\"../../../../assets/fonts/FedraSansStd-Medium.woff\\") format(\\"woff\\"), url(\\"../../../../assets/fonts/FedraSansStd-Medium.ttf\\") format(\\"truetype\\");
  font-weight: 400;
  font-style: normal; }

.dnb-tabs {
  font-family: var(--font-family-book);
  font-size: 1rem;
  font-weight: 300;
  font-style: normal;
  line-height: 1.5em; }
  .dnb-tabs *,
  .dnb-tabs ::before,
  .dnb-tabs ::after {
    background-repeat: no-repeat;
    box-sizing: border-box; }
  .dnb-tabs ::before,
  .dnb-tabs ::after {
    text-decoration: inherit;
    vertical-align: inherit; }

/*
 * Tabs component
 *
 * ========================================================================== */
.dnb-tabs__tabs {
  position: relative;
  z-index: 1;
  display: flex;
  flex-direction: row;
  align-content: flex-start;
  justify-content: space-between;
  align-items: center;
  flex-wrap: nowrap; }
  .dnb-tabs__tabs__tablist {
    display: flex;
    flex-direction: row; }
  .dnb-tabs__tabs--left .dnb-tabs__tabs__tablist {
    justify-content: flex-start; }
  .dnb-tabs__tabs--right .dnb-tabs__tabs__tablist {
    justify-content: flex-end; }
  .dnb-tabs__tabs--center .dnb-tabs__tabs__tablist {
    justify-content: center; }
  .dnb-tabs__tabs::after {
    content: '';
    position: absolute;
    z-index: -1;
    left: -50vw;
    right: -50vw;
    bottom: 0;
    width: 200vw;
    height: 1px; }
  .dnb-tabs__tabs .dnb-tablink {
    position: relative;
    z-index: 1;
    margin: 0 0 0 2.5em;
    padding: 0.5em 0;
    background-color: transparent;
    border-bottom: 1px solid transparent;
    cursor: pointer;
    outline: none;
    border: none; }
    .dnb-tabs__tabs .dnb-tablink-title {
      font-size: 1em;
      line-height: 1.5em;
      white-space: nowrap; }
    .dnb-tabs__tabs .dnb-tablink > span[hidden] {
      display: flex;
      flex-direction: column;
      visibility: hidden;
      height: 0;
      overflow: hidden;
      letter-spacing: 2px; }
    .dnb-tabs__tabs .dnb-tablink::after {
      content: '';
      position: absolute;
      z-index: -1;
      left: 0;
      right: 0;
      bottom: 0;
      width: 100%;
      height: 1px; }
  .dnb-tabs__tabs .dnb-tablink:first-child {
    margin-left: 0; }
  @media screen and (max-width: 700px) {
    .dnb-tabs__tabs__tablist {
      display: block; }
    .dnb-tabs__tabs .dnb-tablink {
      margin-top: 0.5em;
      margin-left: 0; } }

.dnb-tabs__content {
  padding: 1rem 0.5rem; }
"
`;
